torch/distributed/nn/jit/templates/remote_module_template.py

```diff
@@ -13,13 +13,47 @@
 
 
 {jit_script_decorator}
-def _remote_forward(module_rref: RRef[module_interface_cls], {arg_types}){arrow_and_return_type}:
+def _remote_forward(
+    module_rref: RRef[module_interface_cls], device: str, {arg_types}){arrow_and_return_type}:
     module = module_rref.local_value()
-    return module.forward({args}, {kwargs})
+    device = torch.device(device)
+
+    if device.type != "cuda":
+        return module.forward({args}, {kwargs})
+
+    # If the module is on a cuda device,
+    # move any CPU tensor in args or kwargs to the same cuda device.
+    # Since torch script does not support generator expression,
+    # have to use tuple concatenation instead of
+    # ``tuple(i.to(device) if isinstance(i, Tensor) else i for i in *args)``.
+    args = ({args},)
+    out_args = ()
+    for arg in args:
+        arg = (arg.to(device),) if isinstance(arg, Tensor) else (arg,)
+        out_args = out_args + arg
+
+    kwargs = {{{kwargs}}}
+    for k, v in kwargs.items():
+        if isinstance(v, Tensor):
+            kwargs[k] = kwargs[k].to(device)
+
+    # Since only CPU tensors are allowed to send over wire,
+    # need to move any GPU tensor to CPU in the output.
```

Comment:
Do we get an error currently if we leave the outputs on GPU when the backend is tensorpipe? IIUC, it should be okay to leave the outputs on GPU, and TP agent will ensure that on the caller they will be on the appropriate cuda device for that worker via the global device_maps. Is there any reason we can't use this functionality here?

Reply:
To leave the output on GPU, TensorPipe backend requires passing an explicit device map.

Otherwise, it will throw an error. See:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L1460

Reply:
I see, that's interesting. @mrshenli do you know why the global device maps aren't picked up here, and is that an expected occurence?